docs/spikes/2020-04-how-to-migrate-lb3-components.md

```diff
@@ -0,0 +1,650 @@
+
+4. Contribute custom entities (Application, Installation) to be persisted via
+   CRUD, exposed via REST and possibly further customized by the app.
+   Customization options include which datasource to use, the base path where
+   REST API is exposed (e.g. `/api/apps` and `/api/devices`), additional fields
+   (e.g. `Application.tenantId`) and changes in persistence behavior (e.g. via
+   Operation Hooks).
+
+5. Contribute custom models (Notification) describing shape of data expected by
+   the services (Push service). These models are not backed by any datasource,
+   they are primarily used to describe data fields.
+
+6. Add a custom Operation Hook to given models, with a config option to
+   enable/disable this feature. The list of models can be provided explicitly in
+   the component configuration or obtained dynamically via introspection (e.g.
+   all models having a "belongsTo" relation with the Group model)
+
+7. Add a custom Operation Hook to all app models matching given criteria (have a
+   "belongsTo" relation with the Group model, are listed in the component
+   config), plus a config option to enable/disable this feature.
+
+8. Add new relations, e.g. between an app-provided entity `User` and a
+   component-provided entity `File`. In this variant, the relation is added on
+   small fixed number of models.
+
+9. A model mixing adding new relations (`hasMany ModelEvents`), installing
+   Operation Hooks (to generate model events/audit log entries), adding new
+   Repository APIs (for working with related model events).
+
+   _(The mixin-based design may be difficult to accomplish in LB4, we may want
+   to use introspection and a model setting instead. The trickier part is how to
+   apply changes to models added after the component was mounted.)_
+
+10. For all models with a flag enabled in model settings, setup a custom
+    `afterRemote` hook to modify the HTTP response (e.g. add additional
+    headers).
+
+### REST API
+
+11. Add a new REST API endpoint to every (public) model/entity provided by the
+    app, e.g. `GET /api/{model}/schema` returning JSON schema of the model.
+
+12. Add a new REST API endpoint at a configured path, e.g. `/visualize`
+    returning an HTML page.
+
+13. REST API endpoints providing file upload & download.
+
+14. Add a new local service (e.g. `Ping.ping()`) and expose it via REST API
+    (e.g. `/ping`), allow the user to customize the base path where the API is
+    exposed at (e.g. `/pong`).
+
+15. Intercept REST requests handled by an endpoint contributed by a 3rd-party
+    controller to invoke additional logic before the controller method is called
+    (e.g. automatically create containers when uploading files) and after the
+    controller method returned (e.g. post-process the uploaded file).
+
+16. Add new REST API endpoints using Express `(req, res, next)` style.
+
+### Services (local and remote)
+
+16. Add a new local service (a class providing JS/TS API), e.g. Push service or
```

Comment:
Nitpick: I think the numbering is one off from here down

Reply:
Great catch, thank you!